cabal-install/src/Distribution/Client/CmdRepl.hs

```diff
@@ -324,77 +352,15 @@ data OriginalComponentInfo = OriginalComponentInfo
-                                    --   script responsible for changing to the
-                                    --   correct directory. Only works on GHC geq
-                                    --   7.6, though. 🙁
-              deriving (Show, Eq)
-
-withProject :: ProjectConfig -> Verbosity -> [String]
-            -> IO (ProjectBaseContext, [TargetSelector], IO (), ReplType)
-withProject cliConfig verbosity targetStrings = do
-  baseCtx <- establishProjectBaseContext verbosity cliConfig OtherCommand
-
-  targetSelectors <- either (reportTargetSelectorProblems verbosity) return
-                 =<< readTargetSelectors (localPackages baseCtx) (Just LibKind) targetStrings
-
-  return (baseCtx, targetSelectors, return (), ProjectRepl)
-
-withoutProject :: ProjectConfig -> Verbosity -> [String]
-               -> IO (ProjectBaseContext, [TargetSelector], IO (), ReplType)
-withoutProject config verbosity extraArgs = do
-  unless (null extraArgs) $
-    die' verbosity $ "'repl' doesn't take any extra arguments when outside a project: " ++ unwords extraArgs
-
-  globalTmp <- getTemporaryDirectory
-  tempDir <- createTempDirectory globalTmp "cabal-repl."
-
-  -- We need to create a dummy package that lives in our dummy project.
-  let
-    sourcePackage = SourcePackage
-      { srcpkgPackageId     = pkgId
-      , srcpkgDescription   = genericPackageDescription
-      , srcpkgSource        = LocalUnpackedPackage tempDir
-      , srcpkgDescrOverride = Nothing
-      }
-    genericPackageDescription = emptyGenericPackageDescription
-      & L.packageDescription .~ packageDescription
-      & L.condLibrary        .~ Just (CondNode library [baseDep] [])
-    packageDescription = emptyPackageDescription
-      { package = pkgId
-      , specVersion = CabalSpecV2_2
-      , licenseRaw = Left SPDX.NONE
-      }
-    library = emptyLibrary { libBuildInfo = buildInfo }
-    buildInfo = emptyBuildInfo
-      { targetBuildDepends = [baseDep]
-      , defaultLanguage = Just Haskell2010
-      }
-    baseDep = Dependency "base" anyVersion mainLibSet
-    pkgId = fakePackageId
-
-  writeGenericPackageDescription (tempDir </> "fake-package.cabal") genericPackageDescription
-
-  let ghciScriptPath = tempDir </> "setcwd.ghci"
-  cwd <- getCurrentDirectory
-  writeFile ghciScriptPath (":cd " ++ cwd)
-
-  distDirLayout <- establishDummyDistDirLayout verbosity config tempDir
-  baseCtx <-
-    establishDummyProjectBaseContext
-      verbosity
-      config
-      distDirLayout
-      [SpecificSourcePackage sourcePackage]
-      OtherCommand
-
-  let
-    targetSelectors = [TargetPackage TargetExplicitNamed [pkgId] Nothing]
-    finalizer = handleDoesNotExist () (removeDirectoryRecursive tempDir)
-
-  return (baseCtx, targetSelectors, finalizer, GlobalRepl ghciScriptPath)
+-- Tracks what type of GHCi instance we're creating.
+data ReplType = ProjectRepl | GlobalRepl
+  deriving (Show, Eq)
+
+relativePathBackToCurrentDirectory :: FilePath -> IO FilePath
```

Comment:
not sure what is the intent but have you take a look to

https://github.com/haskell/cabal/blob/4309d0f60089bfe308af63de8671e1723c284c92/cabal-install/src/Distribution/Client/Utils.hs#L222-L230
?

Reply:
makeRelativeToDir seems like what I want. Good catch!

Reply:
When removing `relativePathBackToCurrentDirectory` in favor of `makeRelativeToDir`, I discovered a bug in my code that I'm having a lot of trouble fixing properly. When configuring the project, the script file is relative to the hs-source-dir, but when running the repl, the script file is actually loaded after the repl changes directory, so it needs a different path to the script file.

The final command line used to launch ghc would be easy to fix, but I think it gets constructed somewhere deep in the code that I don't think I have easy access to, and none of the components exposed in replAction seem to help, as far as I've been able to tell.

Reply:
The hacky solution: use a path with a number of "../"' which is max (# path segments of project root) (# path segments of working directory) followed by the absolute path to the script. That way we're guaranteed to go up to the root and can then find the script by it's absolute path. This is also why it seemed to work before, because in all my testing (# path segments of project root) was greater than (# path segments of working directory).

Other possible solution: relax the restriction that hs-source-dirs has to be relative. Maybe we could check that the package-id maches the fake package-id constant and only allow it then?

Reply:
As recently as 3.4.0.0 cabal did accept absolute paths in `hs-source-dirs`

Reply:
See https://github.com/haskell/cabal/issues/6787#issuecomment-991206955. (The plot thickens).